feat(js): Dynamic Actioncable WebSocket URL

audiolion opened the pull request:
### Summary

Allow createWebSocketURL fn to accept a function to generate the websocket URL rather than a string.

Closes #33064 

Please let me know if you would like me to update CHANGELOG.

I believe this would require a need to update the docs to document this functionality, but I am not exactly sure where it would best be placed, any help here is appreciated!

audiolion committed:
feat(js): Dynamic Actioncable WebSocket URL

Allow createWebSocketURL fn to accept a function to generate the websocket URL rather than a string.

kaspth commented on actioncable/app/javascript/action_cable/index.js:
```diff
@@ -30,14 +30,20 @@ export function getConfig(name) {
+    webSocketURL = url()
+  } else {
+    webSocketURL = url
+  }
```
I'm no JS-hound, but is it possible to use a ternary here? Ala `const webSocketURL = (typeof url === 'function') ? url() : url`

We should add a newline to separate the assignment and the if below regardless though.

audiolion replied:
yeah it can definitely be done with a ternary, sometimes people think they are less readable so I didn't do it here, but easy enough to switch. I'll add a space as well.

kaspth reviewed (approved):
cc @javan, will @audiolion have to recompile the Action Cable source for this as well?

audiolion committed:
address pr feedback

change if statement to ternary, use const for consistency, add spacing after ternary expression

kaspth commented:
Crap. Looks like something got screwed up when I merged and you didn't get commit tracking. I've fixed this in 5f7e01c84c

rmacklin commented on actioncable/app/javascript/action_cable/index.js:
```diff
@@ -30,14 +30,16 @@ export function getConfig(name) {
 
 export function createWebSocketURL(url) {
-  if (url && !/^wss?:/i.test(url)) {
+  const webSocketURL = typeof url === 'function' ? url() : url;
```
@kaspth @audiolion I'm not quite following how this change addresses the original issue (https://github.com/rails/rails/issues/33064)

Here, `url` (if it is a function) is being invoked immediately when `createWebSocketURL` is invoked. Thus, calling:
```js
createWebSocketURL(generateURL)
```
with these changes is equivalent to
```js
createWebSocketURL(generateURL())
```
prior to this PR, right?

As I understand it, the original issue was having the created `Consumer`'s `url` property be computed dynamically. In other words, if this PR addressed that issue, then I would expect the new test to pass if it were modified with this change:

```diff
diff --git a/actioncable/test/javascript/src/unit/action_cable_test.js b/actioncable/test/javascript/src/unit/action_cable_test.js
index c9d34abc6d..2b89439484 100644
--- a/actioncable/test/javascript/src/unit/action_cable_test.js
+++ b/actioncable/test/javascript/src/unit/action_cable_test.js
@@ -43,11 +43,16 @@ module("ActionCable", () => {
     })
 
     test("uses function to generate URL", assert => {
+      let dynamicURL = testURL
+
       const generateURL = () => {
-        return testURL
+        return dynamicURL
       }
       const consumer = ActionCable.createConsumer(generateURL)
       assert.equal(consumer.url, testURL)
+
+      dynamicURL = `${testURL}foo`
+      assert.equal(consumer.url, `${testURL}foo`)
     })
   })
 })
```

However, it fails:
```
Expected: "ws://cable.example.com/foo"
Actual: "ws://cable.example.com/"
```

Am I missing something?

(cc @javan)

javan replied:
That sounds right, @rmacklin. For this to work correctly, we should always pass a function to `Consumer`, not a string, and the consumer should call that function whenever it opens a new connection.

audiolion replied:
great point @rmacklin let me correct this.
